lib/eco/reducer.py

```diff
@@ -0,0 +1,153 @@
+        self.lexer = lexer
+        self.grm = grm
+        self.treemanager = TreeManager()
+        self.treemanager.add_parser(parser, lexer, grm.name)
+        self.treemanager.version = 1
+        self.treemanager.last_saved_version = 1
+
+        self.program = program
+        self.test = test
+        self.pos = 0
+        self.lastexception = None
+
+    def reset(self):
+        self.parser.reset()
+        self.treemanager = TreeManager()
+        self.treemanager.add_parser(self.parser, self.lexer, self.grm)
+
+    def auto(self):
+        logging.info("Running auto-mode")
+        # try to reduce snapshots first
+        while True:
+            self.original = self.test[:] # remember original tests
+            self.reduce("snapshot")
+            if self.test == self.original:
+                logging.info("No more snapshot reductions possible.")
+                break
+        # try single reduce next
+        while True:
+            self.original = self.test[:] # remember original tests
+            self.reduce("single")
+            if self.test == self.original:
+                logging.info("No more single reductions possible.")
+                break
+
+    def reduce(self, mode):
+        self.pos = 0
+        logging.info("Pos: {} Len: {} Start {}-reduce".format(self.pos, len(self.test), mode))
+        try:
+            # Run without changes to store exception
+            self.run()
+        except Exception as e:
+            self.lastexception = e
+        while self.pos < len(self.test):
+            self.delete(mode)
+            try:
+                self.run() # check if first run causes error
+                logging.info("Pos: {} Len: {} Fixed".format(self.pos, len(self.test)))
+                self.revert()
+            except Exception as e:
+                if not self.same_exception(e):
+                    logging.info("Pos: {} Len: {} Wrong exception: '{}'".format(self.pos, len(self.test), e))
+                    self.revert()
+                    continue
+                else:
+                    logging.info("Pos: {} Len: {} Still failing: '{}'".format(self.pos, len(self.test), e))
+                    self.save() # save progress
+        self.save()
+
+    def same_exception(self, e):
+        if not self.lastexception:
+            self.lastexception = e
+            return True
+        if type(self.lastexception) is type(e) and \
+                self.lastexception.args == e.args:
+                    return True
+        return False
+
+    def delete(self, mode):
+        self.old = self.test[:] # copy
+        if mode == "single":
+            del self.test[self.pos:self.pos+4] # remove one keystroke
+            if self.pos < len(self.test) and \
+                    self.test[self.pos] == "self.treemanager.undo_snapshot()":
+                        self.test.pop(self.pos)
+        elif mode == "snapshot":
+            del self.test[self.pos:self.pos+81] # remove one snapshot
```

Comment:
Yes it is intentional since each snapshot contains 20 keystrokes where 1 keystroke consists of 4 lines plus an additional `save_snapshot` line. I don't mind moving it into a global variable tho.